{% extends 'base.html' %}

{% block head %}

<script src="https://cdn.jsdelivr.net/npm/mathjax@2/MathJax.js?config=AM_CHTML"></script>

<title>
    LIGYSIS: Help
</title>

{% endblock%}

{% block main %}

<div class="container">
    <div class="row">
        <div class="col">
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h1>
                            HELP
                        </h1>
                    </div>
                    <div class="row">
                        <p>
                            The <b>LIGYSIS</b> web server is a way to access, visualise, and download the results of our pipeline for the analysis of ligand
                            binding sites, <b>LIGYSIS</b>. 
                        </p>
                    </div>
                </div>
            </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h2>
                            Exploring the LIGYSIS dataset
                        </h2>
                    </div>
                    <div class="row">
                        <p>
                            The <b>LIGYSIS</b> dataset is comprised by 64,782 ligand binding sites defined from 435,038 <a href="https://zhanggroup.org/BioLiP/">biologically relevant</a>
                            ligands binding to 25,003 proteins across 104,456 structures deposited on the <a href="https://www.ebi.ac.uk/pdbe/">PDBe</a>. To explore the dataset, users can search
                            for a protein of interest by its UniProt Accession Identifier, or protein name.

                        </p>
                    </div>
                </div>
            </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h2>
                            Submitting a job to the LIGYSIS web server
                        </h2>
                    </div>
                    <div class="row">
                        <p>
                            
                            <p class="mb-2">
                                Additionally, users can submit their own set of structure for analysis to the LIGYSIS Web server. The input to do so is the following:
                            </p>
                            <div class="col">
                                <ul class="mb-0">
                                    <li class="mb-2">
                                        <b>UniProt Accession</b>: the UniProt Accession Identifier of the protein the structures map to. This is necessary if the protein is present
                                        in UniProt. If the structures map to a protein not present in UniProt, this field can be left empty.
                                    </li>
                                    <li class="mb-2">
                                        <b>Structure files</b>: the set of structure files to be analysed. LIGYSIS is compatible with either be in mmCIF (<i>.cif</i>) or DB (<i>P.ent</i>, <i>P.pdb</i>) format. 
                                        However, all files within the set of structures must present the same format. These structures must represent the same protein unit, i.e., a dimer, or monomer,
                                        bound to the ligands of interest. This is required for the correct superposition of the structures. The current limit of structures to be submitted in a job are 50.
                                        <b>LIGYSIS</b> only supports single protein-ligand complexes, i.e., homomultimers are accepted, but not heteromultimers. Files can be selected from the file browser
                                        or dragged and dropped into the designated area.
                                    </li>
                                </ul>
                            </div>
                            <img src="{{ url_for('static', filename='images/job_submission.png') }}" alt="Job submission form">
                        </p>
                    </div>
                </div>
            </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h2>
                            Results Page
                        </h2>
                    </div>
                    <div class="row mb-3">
                        <p class="mb-2">
                            This is the <b>LIGYSIS</b> results page. At the very top, the UniProt accession, entry and protein names can be found,
                            as well as the number of chains, ligands and sites for this <b>LIGYSIS</b> entry. Below, the three main panels: <b>Binding Sites</b> (left),
                            <b>Structure</b> (centre), and <b>Binding Site Residues</b> (right).
                        </p>
                        <img src="{{ url_for('static', filename='images/LIGYSIS_WEB_RESULTS2.png') }}" alt="LIGYSIS results">
                    </div>
                    <div class="row">
                        <h3 id="structurePanelHelp" class="mb-3">
                            The Structure Panel <i>(centre)</i>
                        </h3>
                        <p class="mb-2">
                            This is the main panel. It includes a <a href="https://3dmol.csb.pitt.edu/">3Dmol.js</a> structure viewer and the structure control buttons. There are 
                            13 buttons.
                        </p>
                        <div class="col">
                            <ol class="mb-0">
                                <li class="mb-2">
                                    <b>SLICE</b>: clicking on this button displays/hides the controls (near and far) to take slices of the current view between the two planes.
                                </li>
                                <li class="mb-2">
                                    <b>SURF</b>: clicking on this button switches the visibility of the protein surfaces.
                                </li>
                                <li class="mb-2">
                                    <b>LABEL</b>: clicking on this button switches the visibility of residue labels of a clicked binding site,
                                    displayed protein-ligand contacts or hovered residue.
                                </li>
                                <li class="mb-2">
                                    <b>LIGAND</b>: clicking on this button switches the visibility of ligands (small molecule).
                                </li>
                                <li class="mb-2">
                                    <b>HOH</b>: clicking on this button switches the visibility of water molecules.
                                </li>
                                <li class="mb-2">
                                    <b>CONTACT</b>: clicking on this button switches the visibility of protein-ligand interactions. Ligands and interacting sidechains will be
                                    displayed as sticks, and dashed cylinders will connect them. The cylinders have hover-over information.
                                </li>
                                <li class="mb-2">
                                    <b>Download Current Assembly Contacts</b>: clicking on this button will download a <em>.csv</em> file with the contacts of the current assembly
                                    as calculated by pdbe-arpeggio. This option is unavailable when viewing the superposition.
                                </li>
                                <li class="mb-2">
                                    <b>Save Image</b>: click here to save the current state of the viewer to a <em>.png</em> image.
                                </li>
                                <li class="mb-2">
                                    <b>Spin</b>: click here to save to start/stop a spin animation on the <b>Y</b> axis.
                                </li>
                                <li class="mb-2">
                                    <b>Download ALL Assemblies Contacts</b>: clicking on this button will download a <i>.zip</i> file containing a 
                                    <em>.csv</em> file with the contacts of each independent assembly as calculated by pdbe-arpeggio.
                                </li>
                                <li class="mb-2">
                                    <b>Download Current Assembly</b>: clicking on this button will download a <em>.zip</em> file containing the necessary files to visualise 
                                    the currently visualised assembly in <a href="https://www.cgl.ucsf.edu/chimerax/">ChimeraX</a> (<i>.cxc</i>) or <a href="https://pymol.org/#page-top"
                                    >PyMol</a> (<i>.pml</i>).
                                </li>
                                <li class="mb-2">
                                    <b>Download Superposition</b>: clicking on this button will download a <em>.zip</em> file containing the necessary files to visualise 
                                    the ligand superposition in either ChimeraX or PyMol.
                                </li>
                                <li class="mb-2">
                                    <b>Download ALL Assemblies</b>: clicking on this button will download a <em>.zip</em> file containing a subdirectory for each assembly on this entry
                                    with the necessary files to visualise each assembly in ChimeraX or PyMol.
                                </li>
                            </ol>
                        </div>
                    </div>
                    <div class="row">
                        <h3 id="sitePanelHelp" class="mb-3">
                            The Binding Site Panel <i>(left)</i>
                        </h3>
                        <p class="mb-2">
                            This panel includes an interactive scatter plot generated with <a href="https://www.chartjs.org/">Chart.js</a>, in which
                            each data point represents a defined ligand binding site on the protein. Three utilities can be found below:
                        </p>
                        <div class="col">
                            <ol class="mb-0">
                                <li class="mb-2">
                                    <b>X-axis</b>: this dropdown menu selects the variable to be plotted on the X-axis.
                                </li>
                                <li class="mb-2">
                                    <b>Y-axis</b>: this dropdown menu selects the variable to be plotted on the Y-axis.
                                </li>
                                <li class="mb-2">
                                    <b>Save Image</b>: clicking here saves the current graph to a <em>.png</em> image.
                                </li>
                            </ol>
                        </div>
                        <p class="mb-2">
                            In addition, this panel includes a dynamic table showing the data underlying the graph. Each table row corresponds
                            to a data point in the plot, i.e., to a ligand binding site on the protein. The columns are:
                        </p>
                        <div class="col">
                            <ul class="mb-0">
                                <li class="mb-2">
                                    <b>ID</b>: the unique identifier of the binding site. This ID is <em>arbitrarily assigned</em> by the pipeline,
                                    i.e., lower/higher numbers do not have any meaning.
                                </li>
                                <li class="mb-2">
                                    <b>RSA</b>: the <a href="https://en.wikipedia.org/wiki/Relative_accessible_surface_area">relative solvent accessibility</a> of the binding site. This is a measure of how exposed the binding
                                    site is to the solvent. It is calculated as the <em>average</em> of the relative solvent accessibility of the residues
                                    forming the binding site.
                                </li>
                                <li class="mb-2">
                                    <b>DS</b>: the divergence score of the binding site. This is a measure of how different in sequence is the binding site
                                    across homologues in a multiple sequence alignment (MSA). It is calculated as the <em>average</em> of the divergence score of the residues forming
                                    the binding site. This is a normalised version of the <a href="https://doi.org/10.1002/prot.340110408">Shenkin divergence score</a>.
                                </li>
                                <li class="mb-2">
                                    <b>MES</b>: the missense enrichment score of the binding site. This is a measure of how constrained the binding site is across human homologues.
                                    It is calculated as the <em>average</em> of the missense enrichment score of the residues forming the binding site.
                                    This is an <a href="https://en.wikipedia.org/wiki/Odds_ratio">odds ratio</a> (OR) calculated using the
                                    <a href="https://gnomad.broadinstitute.org/">gnomAD</a> database.
                                    For more details see <a href="https://doi.org/10.1101/127050">MacGowan <em>et al.</em>, 2017</a>,
                                    <a href="https://doi.org/10.1371/journal.pcbi.1009335">Utgés <em>et al.</em>, 2021</a>,
                                    and <a href="https://www.nature.com/articles/s42003-024-06117-5">MacGowan <em>et al.</em>, 2024</a>.
                                </li>
                                <li class="mb-2">
                                    <b>Size</b>: the size of the binding site in amino acids (aa), i.e., how many amino acids form the binding site.
                                </li>
                                <li class="mb-2">
                                    <b>Cluster</b>: the RSA cluster label predicted for the binding site. This label is obtained using an
                                    <a href="https://en.wikipedia.org/wiki/Artificial_neural_network">
                                    artificial neural network</a> (ANN), specifically a <a href="https://en.wikipedia.org/wiki/Feedforward_neural_network#Multilayer_perceptron">
                                    multilayer perceptron</a> (MLP).
                                    The label is predicted from the solvent accessibility profile of the site and is indicative of the likely functional state of the binding site
                                    (see <a href="https://www.nature.com/articles/s42003-024-05970-8">Utgés <em>et al.</em>, 2024</a>). Cluster 1 is the most likely to be functional, while
                                    Cluster 4 is the least likely.
                                </li>
                                <li class="mb-2">
                                    <b>FS</b>: the functional score of the binding site. The score ranges from 0.04-0.52. The higher the score, the more likely 
                                    the binding site is to be functional. It is calculated based on the following formula:<br>
                                    <div class="cmath text-center">
                                        `FS_(i) = sum_(j=1)^4 p_(i_j) f_(j)`
                                    </div><br>
                                    <div class="cmath">
                                        The functional score of binding site `i`, `FS_(i)`, is defined by `p_(i_j)`, which is the probability of binding site `i` belonging to
                                        cluster `j`. `p_(i_j)` is given by the outcome vector resulting from the MLP: `P_i`                                        
                                    </div><br>
                                    <div class="cmath text-center">
                                        `P_i = [p_(i_(C_(1))), p_(i_(C_(2))),  p_(i_(C_(3))), p_(i_(C_(4)))]`
                                    </div><br>
                                    <div class="cmath">
                                        Where `p_(i_(C_(1)))` is the probability of binding site `i` to belong to cluster `C_1`, and so on.<br><br>
                                        Finally, `f_(j)` is the functional score of cluster `j`.The functional score of each cluster is calculated as the
                                        <em>proportion</em> of known functional sites, as annotated in <a href="https://www.uniprot.org/">UniProt</a>, within that cluster.<br><br>
                                    </div>
                                    
                                    <div class="cmath">
                                        The functional scores of the clusters are given by `F = [0.52, 0.18, 0.05, 0.04]` for `C_1`, `C_2`, `C_3`, and `C_4`, respectively. 13,000 human
                                        protein-ligand binding sites were used to calculate these proportions of functional sites within each cluster.

                                        <br><br>
                                        At the moment, this score is purely based on the RSA Cluster label. Future versions will also include evolutionary
                                        divergence, as well as missense variation information.
                                    </div>
                                </li>
                            </ul>
                        </div>
                        <p class="mb-2">
                            This table can be downloaded as a <em>.csv</em> file by clicking on the <b>Download Table</b> button. Information about the columns can be found on hovering
                            on the column names, or in more detail by clicking on the <b>Column Information</b> button, which leads to this <b>Help</b> page.
                        </p>
                    </div>
                    <div class="row">
                        <h3 id="siteResiduesPanelHelp" class="mb-3">
                            The Binding Site Residues Panel <i>(right)</i>
                        </h3>
                        <p class="mb-2">
                            This panel also includes an interactive scatter plot generated with <a href="https://www.chartjs.org/">Chart.js</a>, but now
                            each data point represents a <em>binding site residue</em>. Below this, three utilities can be found:
                        </p>
                        <div class="col">
                            <ol class="mb-0">
                                <li class="mb-2">
                                    <b>X-axis</b>: this dropdown menu allows the user to select the variable to be plotted on the X-axis.
                                </li>
                                <li class="mb-2">
                                    <b>Y-axis</b>: this dropdown menu allows the user to select the variable to be plotted on the Y-axis.
                                </li>
                                <li class="mb-2">
                                    <b>Save Image</b>: click here to save the current graph to a <em>.png</em> image.
                                </li>
                            </ol>
                        </div>
                        <p class="mb-2">
                            In addition, this panel includes a dynamic table showing the data underlying the graph. Each table row corresponds
                            to a data point in the plot, i.e., to a residue forming a binding site on the protein. The columns are:
                        </p>
                        <div class="col">
                            <ul class="mb-0">
                                <li class="mb-2">
                                    <b>UPResNum</b>: the residue number identifier matching the UniProt sequence, or UniProt Residue Number.
                                </li>
                                <li class="mb-2">
                                    <b>MSACol</b>: the multiple sequence alignment (MSA) column index where a residue aligns to.
                                </li>
                                <li class="mb-2">
                                    <b>DS</b>: the divergence score of the alignment column where a residue is aligned. The score used is a
                                    normalised version of the <a href="https://doi.org/10.1002/prot.340110408">Shenkin divergence score</a>.
                                </li>
                                <li class="mb-2">
                                    <b>MES</b>: the missense enrichment score of the alignment column where a residue aligns to.
                                    This is an <a href="https://en.wikipedia.org/wiki/Odds_ratio">odds ratio</a> (OR) calculated using the
                                    <a href="https://gnomad.broadinstitute.org/">gnomAD</a> database.
                                    For more details see <a href="https://doi.org/10.1101/127050">MacGowan <em>et al.</em>, 2017</a>,
                                    <a href="https://doi.org/10.1371/journal.pcbi.1009335">Utgés <em>et al.</em>, 2021</a>,
                                    and <a href="https://www.nature.com/articles/s42003-024-06117-5">MacGowan <em>et al.</em>, 2024</a>.
                                </li>
                                <li class="mb-2">
                                    <b>p</b>: the p-value of the MES (OR), obtained using <a href="https://en.wikipedia.org/wiki/Fisher%27s_exact_test">Fisher's exact test</a>.
                                </li>
                                <li class="mb-2">
                                    <b>AA</b>: the amino acid one letter code of a given residue.
                                </li>
                                <li class="mb-2">
                                    <b>RSA</b>: the relative solvent accessibility of a given residue. This is a normalised version of the
                                    accessible surface area (ASA) calculated by <a href="https://doi.org/10.1002/bip.360221211">DSSP</a>.
                                </li>
                                <li class="mb-2">
                                    <b>SS</b>: the secondary structure element of a given residue as calculated by DSSP.
                                </li>
                            </ul>
                        </div>
                        <p class="mb-2">
                            This table can be downloaded as a <em>.csv</em> file by clicking on the <b>Download Table</b> button. Information about the columns can be found on hovering
                            on the column names, or in more detail by clicking on the <b>Column Information</b> button, which leads to this <b>Help</b> page. Additionally, the multiple
                            sequence alignment (MSA) of the protein can be downloaded as a <em>.sto</em> file by clicking on the <b>Download MSA</b> button.
                        </p>
                    </div>
            <!-- <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h2>
                            Coming soon...
                        </h2>
                    </div>
                    <div class="row">
                        <p>
                            Currently, the <b>LIGYSIS</b> server only allows for the inspection of proteins that are in our database. Future versions will allow users to submit
                            their own set of protein 3D structures for analysis. These structures should be in <span class="key-concept">PDB format</span>, as well as represent the
                            <span class="key-concept">same protein unit</span>, i.e., a dimer, or monomer, or hexamer, etc, bound to the ligands of interest. This is required for the
                            <span class="key-concept">correct superposition</span> of the structures. The pipeline will then be run on the user's structures, and the results will be made
                            available for download. <s>Template Jupyter Notebooks, as well as the code for this pipeline can be found <a href="https://github.com/JavierSanchez-Utges/fragsys_custom">here</a>.</s>
                        </p>
                        
                        <p>
                            <br>
                            Ligands and contacts visualisation are not yet implemented in this version. This is at the top of our list, so users can explore a Binding
                            Site of interest and explore which molecules, their type, the residues they interact with and the kind of interactions they form.
                        </p>
                    </div>
                </div>
            </div> -->
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h2>
                            Code availability
                        </h2>
                    </div>
                    <div class="row">
                        <!-- <p style="font-size: 1.5rem; color:red;">
                            The code for the Pipeline nor the Web Server is not available yet.
                        </p> -->
                        <p>
                            The code for the <b>LIGYSIS</b> pipeline can be found <a href="https://github.com/JavierSanchez-Utges/ligysis_dev/tree/running-arpeggio">here</a>.
                            <br><br>
                            The code for the <b>LIGYSIS</b> web server can be found <a href="https://github.com/JavierSanchez-Utges/ligysis_flask">here</a>.
                            <br><br>
                            The code for the <b>LIGYSIS</b> customised pipeline for user jobs can be found <a href="https://github.com/JavierSanchez-Utges/fragsys_custom">here</a>.
                        </p>
                    </div>
                </div>
            </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h2>
                            Citing LIGYSIS
                        </h2>
                    </div>
                    <div class="row">
                        <div class="col">
                            <p class="pull-left">
                                If you use <b>LIGYSIS</b> web server or the <b>LIGYSIS</b> dataset, please cite the following references:
                                <br><br>
                                <ul>
                                    <li>
                                        Utgés <em>et al.</em> Classification of likely functional class for ligand binding sites identified from fragment screening. <em>Commun Biol</em> <b>7</b>, 320 (2024).
                                        <a href="https://doi.org/10.1038/s42003-024-05970-8">
                                            https://doi.org/10.1038/s42003-024-05970-8
                                        </a>
                                    </li>
                                    <br>
                                    <li>
                                        Utgés, J.S., Barton, G.J. Comparative evaluation of methods for the prediction of protein–ligand binding sites. <em>J Cheminform</em> <b>16</b>, 126 (2024).
                                        <a href="https://doi.org/10.1186/s13321-024-00923-z">
                                            https://doi.org/10.1186/s13321-024-00923-z
                                        </a>
                                    </li>
                                </ul>
                            </p>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div> 
</div>

{% endblock%}