{% extends 'base.html' %}

{% block head %}

<script src="https://cdn.jsdelivr.net/npm/mathjax@2/MathJax.js?config=AM_CHTML"></script>

<title>
    LIGYSIS: Help
</title>

{% endblock%}

{% block main %}

<div class="container">
    <div class="row">
        <div class="col">
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h1>
                            HELP
                        </h1>
                    </div>
                    <div class="row">
                        <p>
                            The <b>LIGYSIS</b> web server is a way to access, visualise, and download the results of our pipeline for the analysis of ligand
                            binding sites, <b>LIGYSIS</b>. 
                        </p>
                    </div>
                </div>
            </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h2>
                            Results page
                        </h2>
                    </div>
                    <div class="row mb-3">
                        <p class="mb-2">
                            This is how the main <b>LIGYSIS</b> results page looks like. It is composed by 3 main panels.
                        </p>

                        <img src="/static/images/ligysis_web_results_panels4.png" alt="LIGYSIS results">
                    </div>
                    <div class="row">
                        <h3 id="structurePanelHelp" class="mb-3">
                            The Structure panel
                        </h3>
                        <p class="mb-2">
                            This is the main panel. It includes a <a href="https://3dmol.csb.pitt.edu/">3DMol.js</a> structure viewer and a few utilities
                            below it in the form of buttons. There are four buttons:
                        </p>
                        <div class="col">
                            <ol class="mb-0">
                                <li class="mb-2">
                                    <b>Surface ON/OFF</b>: clicking on this button switches the visibility of the protein surfaces.
                                </li>
                                <li class="mb-2">
                                    <b>Labels ON/OFF</b>: clicking on this button switches the visibility of the labels of the residues forming
                                    the binding site that is clicked on.
                                </li>
                                <li class="mb-2">
                                    <b>Save Image</b>: click here to save the current state of the viewer to a <em>.png</em> image.
                                </li>
                                <li class="mb-2">
                                    <b>Save Structure</b>: click here to save the structure that is being visualised to a file in <em>.cif</em> format.
                                </li>
                            </ol>
                        </div>
                    </div>
                    <div class="row">
                        <h3 id="sitePanelHelp" class="mb-3">
                            The Binding Site panel
                        </h3>
                        <p class="mb-2">
                            This panel includes an interactive scatter plot generated with <a href="https://www.chartjs.org/">Chart.js</a>, in which
                            each data point represents a defined ligand binding site on the protein. Below this, three utilities can be found:
                        </p>
                        <div class="col">
                            <ol class="mb-0">
                                <li class="mb-2">
                                    <b>X-axis</b>: this dropdown menu allows the user to select the variable to be plotted on the x-axis.
                                </li>
                                <li class="mb-2">
                                    <b>Y-axis</b>: this dropdown menu allows the user to select the variable to be plotted on the y-axis.
                                </li>
                                <li class="mb-2">
                                    <b>Save Image</b>: click here to save the current graph to a <em>.png</em> image.
                                </li>
                            </ol>
                        </div>
                        <p class="mb-2">
                            In addition, this panel includes a dynamic table showing the data underlying the graph. Each table row corresponds
                            to a data point in the plot, i.e., to a ligand binding site on the protein. The columns are:
                        </p>
                        <div class="col">
                            <ul class="mb-0">
                                <li class="mb-2">
                                    <b>ID</b>: the unique identifier of the binding site. This ID is <em>arbitrarily assigned</em> by the pipeline,
                                    i.e., lower/higher numbers do not have any meaning.
                                </li>
                                <li class="mb-2">
                                    <b>RSA</b>: the relative solvent accessibility of the binding site. This is a measure of how exposed the binding
                                    site is to the solvent. It is calculated as the <em>average</em> of the relative solvent accessibility of the residues
                                    forming the binding site.
                                </li>
                                <li class="mb-2">
                                    <b>DS</b>: the divergence score of the binding site. This is a measure of how different in sequence is the binding site
                                    across homologues in a multiple sequence alignment (MSA). It is calculated as the <em>average</em> of the divergence score of the residues forming
                                    the binding site. This is a normalised version of the <a href="https://doi.org/10.1002/prot.340110408">Shenkin divergence score</a>.
                                </li>
                                <li class="mb-2">
                                    <b>MES</b>: the missense enrichment score of the binding site. This is a measure of how constrained the binding site is.
                                    It is calculated as the <em>average</em> of the missense enrichment score of the residues forming the binding site.
                                    This is an <a href="https://en.wikipedia.org/wiki/Odds_ratio">odds ratio</a> (OR) calculated using the
                                    <a href="https://gnomad.broadinstitute.org/">gnomAD</a> database.
                                    For more details see <a href="https://doi.org/10.1101/127050">MacGowan <em>et al.</em>, 2017</a>,
                                    <a href="https://doi.org/10.1371/journal.pcbi.1009335">Utgés <em>et al.</em>, 2021</a>,
                                    and <a href="https://www.nature.com/articles/s42003-024-06117-5">MacGowan <em>et al.</em>, 2024</a>.
                                </li>
                                <li class="mb-2">
                                    <b>Size</b>: the size of the binding site in amino acids (aa), i.e., how many amino acids form the binding site.
                                </li>
                                <li class="mb-2">
                                    <b>Cluster</b>: the RSA cluster label predicted for the binding site. This label is obtained using an
                                    <a href="https://en.wikipedia.org/wiki/Artificial_neural_network">
                                    artificial neural network</a> (ANN), specifically a <a href="https://en.wikipedia.org/wiki/Feedforward_neural_network#Multilayer_perceptron">
                                    multilayer perceptron</a> (MLP).
                                    The label is predicted from the solvent accessibility profile of the site and is indicative of the likely functional state of the binding site
                                    (see <a href="https://www.nature.com/articles/s42003-024-05970-8">Utgés <em>et al.</em>, 2024</a>).
                                </li>
                                <li class="mb-2">
                                    <b>FS</b>: the functional score of the binding site. The score ranges from 0.04-0.58. The higher the score, the more likely 
                                    the binding site is to be functional. It is calculated based on the following formula:<br>
                                    <div class="cmath text-center">
                                        `FS_(i) = sum_(j=1)^4 p_(i_j) f_(j)`
                                    </div><br>
                                    <div class="cmath">
                                        The functional score of binding site `i`, `FS_(i)`, is defined by `p_(i_j)`, which is the probability of binding site `i` belonging to
                                        cluster `j`. `p_(i_j)` is given by the outcome vector resulting from the MLP: `P_i`                                        
                                    </div><br>
                                    <div class="cmath text-center">
                                        `P_i = [p_(i_(C_(1))), p_(i_(C_(2))),  p_(i_(C_(3))), p_(i_(C_(4)))]`
                                    </div><br>
                                    <div class="cmath">
                                        Where `p_(i_(C_(1)))` is the probability of binding site `i` to belong to cluster `C_1`, and so on.<br><br>
                                        Finally, `f_(j)` is the functional score of cluster `j`.The functional score of each cluster is calculated as the
                                        <em>proportion</em> of known functional sites, as annotated in UniProt, within that cluster.<br><br>
                                    </div>
                                    
                                    <div class="cmath">
                                        The functional scores of the clusters are given by `F = [0.58, 0.18, 0.05, 0.04]` for `C_1`, `C_2`, `C_3`, and `C_4`, respectively.
                                        <br><br>
                                        At the moment, this score is purely based on the RSA Cluster label. Future versions will also include evolutionary
                                        divergence, as well as missense variation information.
                                    </div>
                                </li>
                            </ul>
                        </div>
                    </div>
                    <div class="row">
                        <h3 id="siteResiduesPanelHelp" class="mb-3">
                            The Binding Site Residues panel
                        </h3>
                        <p class="mb-2">
                            This panel also includes an interactive scatter plot generated with <a href="https://www.chartjs.org/">Chart.js</a>, but now
                            each data point represents a <em>binding site residue</em>. Below this, three utilities can be found:
                        </p>
                        <div class="col">
                            <ol class="mb-0">
                                <li class="mb-2">
                                    <b>X-axis</b>: this dropdown menu allows the user to select the variable to be plotted on the X-axis.
                                </li>
                                <li class="mb-2">
                                    <b>Y-axis</b>: this dropdown menu allows the user to select the variable to be plotted on the Y-axis.
                                </li>
                                <li class="mb-2">
                                    <b>Save Image</b>: click here to save the current graph to a <em>.png</em> image.
                                </li>
                            </ol>
                        </div>
                        <p class="mb-2">
                            In addition, this panel includes a dynamic table showing the data underlying the graph. Each table row corresponds
                            to a data point in the plot, i.e., to a residue forming a binding site on the protein. The columns are:
                        </p>
                        <div class="col">
                            <ul class="mb-0">
                                <li class="mb-2">
                                    <b>UPResNum</b>: the residue number identifier matching the UniProt sequence, or UniProt Residue Number.
                                </li>
                                <li class="mb-2">
                                    <b>MSACol</b>: the multiple sequence alignment (MSA) column index where a residue aligns to.
                                </li>
                                <li class="mb-2">
                                    <b>DS</b>: the divergence score of the alignment column where a residue is aligned. The score used is a
                                    normalised version of the <a href="https://doi.org/10.1002/prot.340110408">Shenkin divergence score</a>.
                                </li>
                                <li class="mb-2">
                                    <b>MES</b>: the missense enrichment score of the alignment column where a residue aligns to.
                                    This is an <a href="https://en.wikipedia.org/wiki/Odds_ratio">odds ratio</a> (OR) calculated using the
                                    <a href="https://gnomad.broadinstitute.org/">gnomAD</a> database.
                                    For more details see <a href="https://doi.org/10.1101/127050">MacGowan <em>et al.</em>, 2017</a>,
                                    <a href="https://doi.org/10.1371/journal.pcbi.1009335">Utgés <em>et al.</em>, 2021</a>,
                                    and <a href="https://www.nature.com/articles/s42003-024-06117-5">MacGowan <em>et al.</em>, 2024</a>.
                                </li>
                                <li class="mb-2">
                                    <b>p</b>: the p-value of the MES (OR), obtained using <a href="https://en.wikipedia.org/wiki/Fisher%27s_exact_test">Fisher's exact test</a>.
                                </li>
                                <li class="mb-2">
                                    <b>AA</b>: the amino acid one letter code of a given residue.
                                </li>
                                <li class="mb-2">
                                    <b>RSA</b>: the relative solvent accessibility of a given residue. This is a normalised version of the
                                    accessible surface area (ASA) calculated by <a href="https://doi.org/10.1002/bip.360221211">DSSP</a>.
                                </li>
                                <li class="mb-2">
                                    <b>SS</b>: the secondary structure element of a given residue as calculated by DSSP.
                                </li>
                            </ul>
                        </div>
                    </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h2>
                            Coming soon...
                        </h2>
                    </div>
                    <div class="row">
                        <p>
                            Currently, the <b>LIGYSIS</b> server only allows for the inspection of proteins that are in our database. Future versions will allow users to submit
                            their own set of protein 3D structures for analysis. These structures should be in <span class="key-concept">PDB format</span>, as well as represent the
                            <span class="key-concept">same protein unit</span>, i.e., a dimer, or monomer, or hexamer, etc, bound to the ligands of interest. This is required for the
                            <span class="key-concept">correct superposition</span> of the structures. The pipeline will then be run on the user's structures, and the results will be made
                            available for download. <s>Template Jupyter Notebooks, as well as the code for this pipeline can be found <a href="https://github.com/JavierSanchez-Utges/fragsys_custom">here</a>.</s>
                        </p>
                        
                        <p>
                            <br>
                            Ligands and contacts visualisation are not yet implemented in this version. This is at the top of our list, so users can explore a Binding
                            Site of interest and explore which molecules, their type, the residues they interact with and the kind of interactions they form.
                        </p>
                    </div>
                </div>
            </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h2>
                            Code availability
                        </h2>
                    </div>
                    <div class="row">
                        <!-- <p style="font-size: 1.5rem; color:red;">
                            The code for the Pipeline nor the Web Server is not available yet.
                        </p> -->
                        <p>
                            The code for the <b>LIGYSIS</b> pipeline can be found <a href="https://github.com/JavierSanchez-Utges/ligysis_dev/tree/running-arpeggio">here</a>.
                            <br><br>
                            The code for the <b>LIGYSIS</b> web server can be found <a href="https://github.com/JavierSanchez-Utges/ligysis_flask">here</a>.
                            <br><br>
                            The code for the <b>LIGYSIS</b> customised pipeline for user jobs can be found <a href="https://github.com/JavierSanchez-Utges/fragsys_custom">here</a>.
                        </p>
                    </div>
                </div>
            </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h2>
                            Citing LIGYSIS
                        </h2>
                    </div>
                    <div class="row">
                        <div class="col">
                            <p class="pull-left">
                                If you use <b>LIGYSIS</b> web server or the <b>LIGYSIS</b> dataset, please cite the following references:
                                <br><br>
                                <ul>
                                    <li>
                                        Utgés <em>et al.</em> Classification of likely functional class for ligand binding sites identified from fragment screening. <em>Commun Biol</em> <b>7</b>, 320 (2024).
                                        <a href="https://doi.org/10.1038/s42003-024-05970-8">
                                            https://doi.org/10.1038/s42003-024-05970-8
                                        </a>
                                    </li>
                                    <br>
                                    <li>
                                        Utgés, J.S., Barton, G.J. Comparative evaluation of methods for the prediction of protein–ligand binding sites. <em>J Cheminform</em> <b>16</b>, 126 (2024).
                                        <a href="https://doi.org/10.1186/s13321-024-00923-z">
                                            https://doi.org/10.1186/s13321-024-00923-z
                                        </a>
                                    </li>
                                </ul>
                            </p>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div> 
</div>

{% endblock%}