{% extends 'base.html' %}

{% block head %}

<script src="https://cdn.jsdelivr.net/npm/mathjax@2/MathJax.js?config=AM_CHTML"></script>

<title>
    LIGYSIS: Help
</title>

{% endblock%}

{% block main %}

<div class="container">
    <div class="row">
        <div class="col">
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h1>
                            HELP
                        </h1>
                    </div>
                    <div class="row">
                        <p>
                            The <b>LIGYSIS</b> web server is a way to access, visualise, and download the results of our pipeline for the analysis of ligand
                            binding sites, <b>LIGYSIS</b>. 
                        </p>
                    </div>
                </div>
            </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <h2>
                            Results page
                        </h2>
                    </div>
                    <div class="row mb-3">
                        <p class="mb-2">
                            This is how the main <b>LIGYSIS</b> results page looks like. It is composed by 3 main panels.
                        </p>

                        <img src="/static/images/ligysis_web_results_panels4.png" alt="LIGYSIS results">
                    </div>
                    <div class="row">
                        <h3 id="structurePanelHelp" class="mb-3">
                            The Structure panel
                        </h3>
                        <p class="mb-2">
                            This is the main panel. It includes a <a href="https://3dmol.csb.pitt.edu/">3DMol.js</a> structure viewer and a few utilities
                            below it in the form of buttons. There are four buttons:
                        </p>
                        <div class="col">
                            <ol class="mb-0">
                                <li class="mb-2">
                                    <b>Surface ON/OFF</b>: clicking on this button switches the visibility of the protein surfaces.
                                </li>
                                <li class="mb-2">
                                    <b>Labels ON/OFF</b>: clicking on this button switches the visibility of the labels of the residues forming
                                    the binding site that is clicked on.
                                </li>
                                <li class="mb-2">
                                    <b>Save Image</b>: click here to save the current state of the viewer to a <em>.png</em> image.
                                </li>
                                <li class="mb-2">
                                    <b>Save Structure</b>: click here to save the structure that is being visualised to a file in <em>.cif</em> format.
                                </li>
                            </ol>
                        </div>
                    </div>
                    <div class="row">
                        <h3 id="sitePanelHelp" class="mb-3">
                            The Binding Site panel
                        </h3>
                        <p class="mb-2">
                            This panel includes an interactive scatter plot generated with <a href="https://www.chartjs.org/">Chart.js</a>, in which
                            each data point represents a defined ligand binding site on the protein. Below this, three utilities can be found:
                        </p>
                        <div class="col">
                            <ol class="mb-0">
                                <li class="mb-2">
                                    <b>X-axis</b>: this dropdown menu allows the user to select the variable to be plotted on the x-axis.
                                </li>
                                <li class="mb-2">
                                    <b>Y-axis</b>: this dropdown menu allows the user to select the variable to be plotted on the y-axis.
                                </li>
                                <li class="mb-2">
                                    <b>Save Image</b>: click here to save the current graph to a <em>.png</em> image.
                                </li>
                            </ol>
                        </div>
                        <p class="mb-2">
                            In addition, this panel includes a dynamic table showing the data underlying the graph. Each table row corresponds
                            to a data point in the plot, i.e., to a ligand binding site on the protein. The columns are:
                        </p>
                        <div class="col">
                            <ul class="mb-0">
                                <li class="mb-2">
                                    <b>ID</b>: the unique identifier of the binding site. This ID is <em>arbitrarily assigned</em> by the pipeline,
                                    i.e., lower/higher numbers do not have any meaning.
                                </li>
                                <li class="mb-2">
                                    <b>RSA</b>: the relative solvent accessibility of the binding site. This is a measure of how exposed the binding
                                    site is to the solvent. It is calculated as the <em>average</em> of the relative solvent accessibility of the residues
                                    forming the binding site.
                                </li>
                                <li class="mb-2">
                                    <b>DS</b>: the divergence score of the binding site. This is a measure of how different in sequence is the binding site
                                    across homologues in a multiple sequence alignment (MSA). It is calculated as the <em>average</em> of the divergence score of the residues forming
                                    the binding site. This is a normalised version of the <a href="https://doi.org/10.1002/prot.340110408">Shenkin divergence score</a>.
                                </li>
                                <li class="mb-2">
                                    <b>MES</b>: the missense enrichment score of the binding site. This is a measure of how constrained the binding site is.
                                    It is calculated as the <em>average</em> of the missense enrichment score of the residues forming the binding site.
                                    This is an <a href="https://en.wikipedia.org/wiki/Odds_ratio">odds ratio</a> (OR) calculated using the
                                    <a href="https://gnomad.broadinstitute.org/">gnomAD</a> database.
                                    For more details see <a href="https://doi.org/10.1101/127050">MacGowan <em>et al.</em>, 2017</a>,
                                    <a href="https://doi.org/10.1371/journal.pcbi.1009335">Utgés <em>et al.</em>, 2021</a>,
                                    and <a href="https://doi.org/10.21203/rs.3.rs-3160340/v1">MacGowan <em>et al.</em>, 2023</a>.
                                </li>
                                <li class="mb-2">
                                    <b>Size</b>: the size of the binding site in amino acids (aa), i.e., how many amino acids form the binding site.
                                </li>
                                <li class="mb-2">
                                    <b>Cluster</b>: the RSA cluster label predicted for the binding site. This label is obtained using an
                                    <a href="https://en.wikipedia.org/wiki/Artificial_neural_network">
                                    artificial neural network</a> (ANN), specifically a <a href="https://en.wikipedia.org/wiki/Feedforward_neural_network#Multilayer_perceptron">
                                    multilayer perceptron</a> (MLP).
                                    The label is predicted from the solvent accessibility profile of the site and is indicative of the likely functional state of the binding site
                                    (see <a href="https://doi.org/10.21203/rs.3.rs-3185838/v1">Utgés <em>et al.</em>, 2023</a>).
                                </li>
                                <li class="mb-2">
                                    <b>FS</b>: the functional score of the binding site. The score ranges from 0.04-0.58. The higher the score, the more likely 
                                    the binding site is to be functional. It is calculated based on the following formula:<br>
                                    <div class="cmath text-center">
                                        `FS_(i) = sum_(j=1)^4 p_(i_j) f_(j)`
                                    </div><br>
                                    <div class="cmath">
                                        Where `FS_(i)` is the functional score of binding site `i`, `p_(i_j)` is the probability of the binding site `i` to belong to
                                        cluster `j`, and `f_(j)` is the functional score of cluster `j`. The functional score of each cluster is calculated as the
                                        <em>proportion</em> of known functional sites, as annotated in UniProt, within that cluster.<br><br>
                                        The functional scores of the clusters are: `f = [0.58, 0.18, 0.05, 0.04]` for `C_1`, `C_2`, `C_3`, and `C_4`, respectively.
                                        <br><br>
                                        At the moment, this score is purely based on the RSA Cluster label. Future versions will also include evolutionary
                                        divergence, as well as missense variation information.

                                    </div>
                                    
                                </li>
                            </ul>
                        </div>
                    </div>

                                    
                                    

                                
                        <h3 id="siteResiduesPanelHelp" class="mb-1">
                            The Binding Site Residues panel
                        </h3>
                                <li id="an_shenkSiteHelp" class="mb-1">
                                    The binding site residues panel:
                                </li>
                                <li id="n_ressSiteHelp" class="mb-1">
                                    The size of the binding site in amino acids (aa), i.e., how many amino acids form the binding site.
                                    This number represents the unique residues that bind the ligands grouped together based on their interaction
                                    fingerprints.
                                </li>

                                
                            <!-- </ol>
                        </div> -->
                        
                        <p>
                            Second paragraph here.
                        </p>
                    </div>
                </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <strong>
                            Coming soon...
                        </strong>
                    </div>
                    <div class="row">
                        <p>
                            Currently, the <b>LIGYSIS</b> server only allows for the inspection of proteins that are in our database. Future versions will allow users to submit
                            their own set of protein 3D structures for analysis. These structures should be in <span class="key-concept">PDB format</span>, as well as represent the
                            <span class="key-concept">same protein unit</span>, i.e., a dimer, or monomer, or hexamer, etc, bound to the ligands of interest. This is required for the
                            <span class="key-concept">correct superposition</span> of the structures. The pipeline will then be run on the user's structures, and the results will be made
                            available for download. Template Jupyter Notebooks, as well as the code for this pipeline can be found <a href="https://github.com/JavierSanchez-Utges/fragsys_custom">here</a>.
                        </p>
                    </div>
                </div>
            </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <strong>
                            Code availability
                        </strong>
                    </div>
                    <div class="row">
                        <p>
                            The code for the <b>LIGYSIS</b> pipeline can be found <a href="https://github.com/JavierSanchez-Utges/ligysis_dev">here</a>.<br>
                            The code for the <b>LIGYSIS</b> web server can be found <a href="https://github.com/JavierSanchez-Utges/ligysis_flask">here</a>.
                        </p>
                    </div>
                </div>
            </div>
            <div class="row mb-3">
                <div class="col">
                    <div class="row mb-2">
                        <strong>
                            Citing LIGYSIS
                        </strong>
                    </div>
                    <div class="row">
                        <p class="pull-left">
                            There is not yet a publication for the <b>LIGYSIS</b> web server. However,
                            this service explores the results obtained with our method, which can be cited:
                            <br><br>
                            J. S. Utgés, et al. Classification of likely functional state for ligand
                            binding sites identified from fragment screening, 24 July 2023, PREPRINT (Version 1), available at
                            <a href="https://doi.org/10.21203/rs.3.rs-3185838/v1">
                                Research Square.
                            </a>
                        </p>
                    </div>
                </div>
            </div>
        </div>
    </div> 
</div>

{% endblock%}